ntp-udp/src/lib.rs

```diff
@@ -1,219 +1,424 @@
+
+        use crate::interface_name::sockaddr_storage_to_socket_addr;
+
+        use super::cerr;
+
+        pub(crate) enum MessageQueue {
+            Normal,
+            Error,
         }
 
-        (libc::SOL_IP, libc::IP_RECVERR) | (libc::SOL_IPV6, libc::IPV6_RECVERR) => {
-            // this is part of how timestamps are reported.
-            let error = unsafe {
-                let ptr = libc::CMSG_DATA(msg) as *const libc::sock_extended_err;
-                std::ptr::read_unaligned(ptr)
+        pub(crate) fn receive_message<'a>(
+            socket: &std::net::UdpSocket,
+            packet_buf: &mut [u8],
+            control_buf: &'a mut [u8],
+            queue: MessageQueue,
+        ) -> std::io::Result<(
+            libc::c_int,
+            impl Iterator<Item = ControlMessage> + 'a,
+            Option<SocketAddr>,
+        )> {
+            let mut buf_slice = IoSliceMut::new(packet_buf);
+            let mut addr = zeroed_sockaddr_storage();
+
+            let mut mhdr = libc::msghdr {
+                msg_control: control_buf.as_mut_ptr().cast::<libc::c_void>(),
+                msg_controllen: control_buf.len(),
+                msg_iov: (&mut buf_slice as *mut IoSliceMut).cast::<libc::iovec>(),
+                msg_iovlen: 1,
+                msg_flags: 0,
+                msg_name: (&mut addr as *mut libc::sockaddr_storage).cast::<libc::c_void>(),
+                msg_namelen: std::mem::size_of::<libc::sockaddr_storage>() as u32,
             };
 
-            ControlMessage::ReceiveError(error)
-        }
-        _ => ControlMessage::Other(*msg),
-    })
-}
+            let receive_flags = match queue {
+                MessageQueue::Normal => 0,
+                MessageQueue::Error => libc::MSG_ERRQUEUE,
+            };
 
-fn raw_control_messages(message_header: &libc::msghdr) -> impl Iterator<Item = &libc::cmsghdr> {
-    let mut cmsg = unsafe { libc::CMSG_FIRSTHDR(message_header).as_ref() };
+            // Safety:
+            // We have a mutable reference to the control buffer for the duration of the
+            // call, and controllen is also set to it's length.
+            // IoSliceMut is ABI compatible with iovec, and we only have 1 which matches iovlen
+            // msg_name is initialized to point to an owned sockaddr_storage and
+            // msg_namelen is the size of sockaddr_storage
+            // If one of the buffers is too small, recvmsg cuts of data at appropriate boundary
+            let sent_bytes = loop {
+                match cerr(
+                    unsafe { libc::recvmsg(socket.as_raw_fd(), &mut mhdr, receive_flags) } as _,
+                ) {
+                    Err(e) if std::io::ErrorKind::Interrupted == e.kind() => {
+                        // retry when the recv was interrupted
+                        continue;
+                    }
+
+                    other => break other,
+                }
+            }?;
+
+            if mhdr.msg_flags & libc::MSG_TRUNC > 0 {
```

Comment:
Is it valid to compare `msg_flags` if `recvmsg()` returned anything other than `std::io::ErrorKind::Interrupted`? It feels like we should only run these blocks if `sent_bytes` is `Ok`?

Reply:
msg_flags is only guaranteed to be valid when recvmsg returns no error. However, this is guaranteed by the question mark operator on line 153.

Reply:
Ahh -- sorry, I missed the question mark. Might consider moving the question mark to the `break other` expression? Honestly not sure if that would be an improvement.

Reply:
Hmm, you have a point that this was getting somewhat confusing to read. Decided to resolve that by refactoring the loop back out into its own function, which makes the question of where to put the question mark moot.

Reply:
Do you agree this is an improvement @djc?

Reply:
IMO it's better to keep the loop inline, just adding indirection (which also means you have to scroll down to be able to get the full picture) doesn't make it easier. There are probably other ways to make it more obvious (for example, you could just explicit arms for the remaining `Err(_)` and `Ok(_)` cases).

Reply:
done